src/SpecFetcher.php

```diff
@@ -0,0 +1,240 @@
+   *
+   * @var \Drupal\Core\File\FileSystemInterface
+   */
+  protected $fileSystem;
+
+  /**
+   * GuzzleHttp\ClientInterface definition.
+   *
+   * @var \GuzzleHttp\ClientInterface
+   */
+  protected $httpClient;
+
+  /**
+   * The entity type manager.
+   *
+   * @var \Drupal\Core\Entity\EntityTypeManagerInterface
+   */
+  protected $entityTypeManager;
+
+  /**
+   * The logger.
+   *
+   * @var \Psr\Log\LoggerInterface
+   */
+  private $logger;
+
+  /**
+   * Constructs a new SpecFetcher.
+   *
+   * @param \Drupal\Core\File\FileSystemInterface $file_system
+   *   The file_system service.
+   * @param \GuzzleHttp\ClientInterface $http_client
+   *   The http_client service.
+   * @param \Drupal\Core\Entity\EntityTypeManagerInterface $entityTypeManager
+   *   The entity type manager.
+   * @param \Drupal\Core\Messenger\MessengerInterface $messenger
+   *   The messenger service.
+   * @param \Psr\Log\LoggerInterface $logger
+   *   The logger service.
+   */
+  public function __construct(FileSystemInterface $file_system, ClientInterface $http_client, EntityTypeManagerInterface $entityTypeManager, MessengerInterface $messenger, LoggerInterface $logger) {
+    $this->fileSystem = $file_system;
+    $this->httpClient = $http_client;
+    $this->entityTypeManager = $entityTypeManager;
+    $this->messenger = $messenger;
+    $this->logger = $logger;
+  }
+
+  /**
+   * {@inheritdoc}
+   */
+  public function fetchSpec(ApiDocInterface $apidoc): bool {
+    $spec_value = $apidoc->get('spec')->isEmpty() ? [] : $apidoc->get('spec')->getValue()[0];
+
+    // If "spec_file_source" uses URL, grab file from "file_link" and save it
+    // into the "spec" file field. The file_link field should already have
+    // validated that a valid file exists at that URL.
+    if ($apidoc->get('spec_file_source')->value === ApiDocInterface::SPEC_AS_URL) {
+
+      // If the file_link field is empty, return without changes.
+      // TODO: The file link shouldn't be empty. Consider throwing an error.
+      if ($apidoc->get('file_link')->isEmpty()) {
+        return FALSE;
+      }
+
+      $source_uri = $apidoc->get('file_link')->getValue()[0]['uri'];
+      $source_uri = Url::fromUri($source_uri, ['absolute' => TRUE])->toString();
+      $request = new Request('GET', $source_uri);
+      $options = [
+        'exceptions' => TRUE,
+        'allow_redirects' => [
+          'strict' => TRUE,
+        ],
+      ];
+
+      // Generate conditional GET header.
+      if (!$apidoc->get('fetched_timestamp')->isEmpty()) {
+        $request = $request->withAddedHeader('If-Modified-Since', gmdate(DateTimePlus::RFC7231, $apidoc->get('fetched_timestamp')->value));
+      }
+
+      try {
+        $response = $this->httpClient->send($request, $options);
+
+        // In case of a 304 Not Modified there are no changes, but update
+        // last fetched timestamp.
+        if ($response->getStatusCode() === 304) {
+          $apidoc->set('fetched_timestamp', time());
```

Comment:
https://github.com/apigee/apigee-edge-drupal/pull/191/files#r284210698

![image](https://user-images.githubusercontent.com/1755573/58001351-820fa280-7adb-11e9-837d-5a594d109351.png)

Reply:
Interesting use case, that could be changed if it is needed in testing.